signature/jws/signer.go

```diff
@@ -0,0 +1,161 @@
+}
+
+// NewSignerWithCertificateChain creates a signer with the specified signing method and
+// a signing key bundled with a certificate chain.
+// The relation of the provided siging key and its certificate chain is not verified,
+// and should be verified by the caller.
+func NewSignerWithCertificateChain(method jwt.SigningMethod, key crypto.PrivateKey, certChain []*x509.Certificate) (*Signer, error) {
+	if method == nil {
+		return nil, errors.New("nil signing method")
+	}
+	if key == nil {
+		return nil, errors.New("nil signing key")
+	}
+	if len(certChain) == 0 {
+		return nil, errors.New("missing signer certificate chain")
+	}
+
+	rawCerts := make([][]byte, 0, len(certChain))
+	for _, cert := range certChain {
+		rawCerts = append(rawCerts, cert.Raw)
+	}
+	return &Signer{
+		method:    method,
+		key:       key,
+		certChain: rawCerts,
+	}, nil
+}
+
+// Sign signs the artifact described by its descriptor, and returns the signature.
+func (s *Signer) Sign(ctx context.Context, desc notation.Descriptor, opts notation.SignOptions) ([]byte, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	// generate JWT
+	payload := packPayload(desc, opts)
+	if err := payload.Valid(); err != nil {
+		return nil, err
+	}
+	token := &jwt.Token{
+		Header: map[string]interface{}{
+			"alg": s.method.Alg(),
+			"cty": MediaTypeNotationPayload,
+			"crit": []string{
+				"cty",
+			},
+		},
+		Claims: payload,
+		Method: s.method,
+	}
+	compact, err := token.SignedString(s.key)
+	if err != nil {
+		return nil, err
+	}
+
+	// generate unprotected header
+	header := unprotectedHeader{
+		CertChain: s.certChain,
+	}
+
+	// timestamp JWT
+	sig, err := jwsutil.ParseCompact(compact)
+	if err != nil {
+		return nil, err
+	}
+	if s.TSA != nil {
+		token, err := s.timestamp(ctx, sig.Signature.Signature)
+		if err != nil {
+			return nil, fmt.Errorf("timestamp failed: %w", err)
+		}
+		header.TimeStampToken = token
+	}
+
+	// finalize unprotected header
+	sig.Unprotected, err = json.Marshal(header)
+	if err != nil {
+		return nil, err
+	}
+
+	// encode in flatten JWS JSON serialization
+	return json.Marshal(sig)
+}
+
+// timestamp sends a request to the TSA for timestamping the signature.
+func (s *Signer) timestamp(ctx context.Context, sig string) ([]byte, error) {
+	// timestamp the signature
+	decodedSig, err := base64.RawURLEncoding.DecodeString(sig)
+	if err != nil {
+		return nil, err
+	}
+	req, err := timestamp.NewRequestFromBytes(decodedSig)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := s.TSA.Timestamp(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	if status := resp.Status; status.Status != pki.StatusGranted {
+		return nil, fmt.Errorf("tsa: %d: %v", status.Status, status.StatusString)
+	}
+	tokenBytes := resp.TokenBytes()
+
+	// verify the timestamp signature
+	if _, err := verifyTimestamp(decodedSig, tokenBytes, s.TSAVerifyOptions); err != nil {
+		return nil, err
```

Comment:
Good point. Let me consider refactoring it.

Reply:
@priteshbandi After taking a look into it, I don't think we need those validations. It is simply because we don't have the request stored in the artifact signature. Even if we store the request in the artifact signature, it can be simply replaced by the adversary since the request is the unprotected header and is not signed.

Reply:
We do need to add more validation for timestamp response to avoid MitM, replay, etc. attacks and also to be a compliant client. The validations are more from timestamp security and correctness perspective. If the timestamp signature is bad then it is of no use for users.

Haven't gone through complete rfc3161 but here are some notable mentions:
* MitM and replay attack: https://datatracker.ietf.org/doc/html/rfc3161#section-4 point 4 and 5
* Policy: https://datatracker.ietf.org/doc/html/rfc3161#section-2.2
* nonce https://datatracker.ietf.org/doc/html/rfc3161#section-2.4.1
* PKIStatus,PKIFailureInfo https://datatracker.ietf.org/doc/html/rfc3161#section-2.4.2

Reply:
Let's address this after this PR.